File: src/Services/Ideas/NetMusicStore.Services.Ideas.Domain/Songs/Song.cs
using NetMusicStore.Services.Ideas.Domain.Shared.ValueObjects;
using NetMusicStore.Services.Ideas.Domain.Songs.Events;
using NetMusicStore.Services.Shared;
using NetMusicStore.Services.Shared.ValueObjects;

namespace NetMusicStore.Services.Ideas.Domain.Songs
{
    public class Song : AggregateRoot
    {
        public string Id { get; private set; }
        public string Name { get; private set; }
        public string Description { get; private set; }
        public double Duration { get; private set; }
        public string DurationFormatted { get; private set; }
        public bool Draft { get; private set; }
        public Author Author { get; private set; }

        private List<SongFile> _files;
        public IReadOnlyCollection<SongFile> Files => _files.AsReadOnly();
        public Audit Audit { get; private set; }
        public ESongStatus Status { get; set; }

        private Song(string name, string description, double duration, string durationFormatted, Author author)
        {
            Name = name;
            Description = description;
            Duration = duration;
            DurationFormatted = durationFormatted;
            _files = new List<SongFile>();
            Audit = Audit.Create(author.UserName);
            Status = ESongStatus.Pending;

            AddSongCreatedDomainEvent(Id, name, durationFormatted, Draft, author.UserName);
        }

        private void AddSongCreatedDomainEvent(string id, string name, string durationFormatted, bool isDraft, string author)
        {
            var songCreated = new SongCreatedDomainEvent(id, name, durationFormatted, isDraft, author);

            AddDomainEvent(songCreated);
        }

        public static Song Create(string name, string description, double duration, string durationFormatted, Author author)
        {
            return new Song(name, description, duration, durationFormatted, author);
        }

        public void MarkAsDraft()
        {
            if (Status != ESongStatus.Pending)
                throw new SongDomainException($"Song cannot be marked as draft. Status: {Status.Name}");

            Draft = true;
        }

        public void Promote()
        {
            if (Status != ESongStatus.Pending)
                StatusChangeException(ESongStatus.Promoted);
            if (Duration < 3)
                throw new SongDomainException("A song cannot be promoted if the duration is less than 4 minutes");

            Status = ESongStatus.Promoted;

            AddDomainEvent(new SongPromotedDomainEvent(Id));
        }

        public void Release()
        {
            if (Status != ESongStatus.Promoted)
                StatusChangeException(ESongStatus.Released);

            Status = ESongStatus.Released;

            AddDomainEvent(new SongReleasedDomainEvent(Id));
        }

        public void Cancel() {
            if (Status == ESongStatus.Promoted || Status == ESongStatus.Released || Status == ESongStatus.Canceled)
                StatusChangeException(ESongStatus.Canceled);

            Status = ESongStatus.Canceled;

            AddDomainEvent(new SongCanceledDomainEvent(Id));
        }

        public void AddFile(string name, string path)
        {
            if (ExistsFile(name))
                throw new SongDomainException($"File {name} exists.");

        }

        public void RemoveFile(string name)
        {
            if (!ExistsFile(name))
                throw new SongDomainException($"File {name} does not exists.");
            
           
        }

        private bool ExistsFile(string name)
        {
            return false;
        }

        private void StatusChangeException(ESongStatus songStatusToChange)
        {
            throw new SongDomainException($"Is not possible to change the order status from {Status.Name} to {songStatusToChange.Name}.");
        }
    }
}

File: src/BackOffice/server/src/services/albums/domain/AlbumAggregate/Song.ts
import { AggregateRoot, Audit } from '../../../shared/domain';
import Author from './Author';
import FileReference from './File';
import SongDomainException from './SongDomainException';

export enum ESongStatus {
  Pending,
  InLab,
  Canceled,
  Closed
}

export const ESongStatusTranslate = {
  [ESongStatus.Pending]: 'Pending',
  [ESongStatus.InLab]: 'InLab',
  [ESongStatus.Canceled]: 'Canceled',
  [ESongStatus.Closed]: 'Closed'
};

class Song extends AggregateRoot {
  private _id: string;
  private _name: string;
  private _description: string;
  private _registerDate: Date;
  private _duration: number;
  private _author: Author;
  private _tags: string[];
  private _files: FileReference[];
  private _audit: Audit;
  private _status: ESongStatus;

  public get id(): string {
    return this._id;
  }

  public get name(): string {
    return this._name;
  }

  public get description(): string {
    return this._description;
  }

  public get registerDate(): Date {
    return this._registerDate;
  }

  public get duration(): number {
    return this._duration;
  }

  public get author(): Author {
    return this._author;
  }

  public get tags(): string[] {
    return this._tags;
  }

  public get files(): FileReference[] {
    return this._files;
  }

  public get audit(): Audit {
    return this._audit;
  }

  public get status(): ESongStatus {
    return this._status;
  }

  private constructor(
    id: string,
    name: string,
    description: string,
    duration: number,
    durationFormatted: string,
    tags: string[],
    author: Author
  ) {
    super();

    const registerDate = new Date();

    this._id = id;
    this._name = name;
    this._description = description;
    this._registerDate = registerDate;
    this._duration = duration;
    this._tags = tags;
    this._author = author;
    this._files = [];
    this._audit = Audit.Create(this._author.authorId);
    this._status = ESongStatus.Pending;
  }

  public static Create(
    id: string,
    name: string,
    description: string,
    duration: number,
    durationFormatted: string,
    tags: string[],
    author: Author
  ): Song {
    return new Song(
      id,
      name,
      description,
      duration,
      durationFormatted,
      tags,
      author
    );
  }

  public Promote() {
    if (this._status !== ESongStatus.Pending)
      this.StatusChangeException(ESongStatus.InLab);

    this._status = ESongStatus.InLab;
    this._audit = this._audit.Update(this._author.authorId);
  }

  private StatusChangeException(songStatusToChange: ESongStatus) {
    throw new SongDomainException(
      `Is not possible to change the album status from ${
        ESongStatusTranslate[this._status]
      } to ${ESongStatusTranslate[songStatusToChange]}`
    );
  }

  public Cancel() {
    if (
      this._status !== ESongStatus.InLab &&
      this._status !== ESongStatus.Pending
    )
      this.StatusChangeException(ESongStatus.Canceled);

    this._status = ESongStatus.Canceled;
    this._audit = this._audit.Update(this._author.authorId);
  }

  private ExistsTag(tag: string): boolean {
    return (
      this._tags.find((p) => p.toLowerCase() === tag.toLowerCase()) !== '' ??
      false
    );
  }

  public AddTag(tag: string): void {
    if (this.ExistsTag(tag)) throw new SongDomainException(`Tag ${tag} exists`);

    this._tags.push(tag);
  }

  public RemoveTag(tag: string): void {
    if (!this.ExistsTag(tag))
      throw new SongDomainException(`Tag ${tag} does not exists`);

    this._tags.splice(
      this._tags.findIndex((p) => p.toLowerCase() === tag.toLowerCase()),
      1
    );
  }

  public AddFile(name: string, path: string): void {
    if (this.ExistsFile(path))
      throw new SongDomainException(`File ${path} exists`);

    this._files.push(FileReference.Create(name, path));
  }

  public RemoveFile(file: FileReference): void {
    const name = file.name;

    if (!this.ExistsFile(file.name))
      throw new SongDomainException(`File ${name} does not exists`);

    this._files.splice(
      this._files.findIndex(
        (p) => p.path.toLowerCase() === file.path.toLowerCase()
      ),
      1
    );
  }

  private ExistsFile(path: string): boolean {
    return (
      this._files.find((p) => p.path.toLowerCase() === path.toLowerCase()) !==
        null ?? false
    );
  }
}

export default Song;
